File: src/main/java/io/github/intellij/dlanguage/stubs/types/ModuleDeclarationStubElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import io.github.intellij.dlanguage.psi.named.DlangModuleDeclaration;
import io.github.intellij.dlanguage.psi.impl.named.DlangModuleDeclarationImpl;
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes;
import io.github.intellij.dlanguage.stubs.DlangModuleDeclarationStub;
import java.io.IOException;
import org.jetbrains.annotations.NotNull;

public class ModuleDeclarationStubElementType extends DNamedStubElementType<DlangModuleDeclarationStub, DlangModuleDeclaration> {
    public ModuleDeclarationStubElementType(final String debugName) {
        super(debugName);
    }

    @Override
    public DlangModuleDeclaration createPsi(@NotNull final DlangModuleDeclarationStub stub) {
        return new DlangModuleDeclarationImpl(stub, this);
    }

    @NotNull
    @Override
    public DlangModuleDeclarationStub createStub(@NotNull final DlangModuleDeclaration psi, final StubElement parentStub) {
        return new DlangModuleDeclarationStub(parentStub, this, psi.getName(), psi.getAttributes());
    }

    @Override
    public void serialize(@NotNull final DlangModuleDeclarationStub stub, @NotNull final StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
        stub.getAttributes().write(dataStream);
    }

    @NotNull
    @Override
    public DlangModuleDeclarationStub deserialize(@NotNull final StubInputStream dataStream, final StubElement parentStub) throws IOException {
        return new DlangModuleDeclarationStub(parentStub, this, dataStream.readName(),
            DAttributes.Companion.read(dataStream));
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/ModuleDeclarationStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.psi.impl.named.DlangModuleDeclarationImpl
import io.github.intellij.dlanguage.psi.named.DLanguageModuleDeclaration
import io.github.intellij.dlanguage.resolve.processors.parameters.DAttributes.Companion.read
import io.github.intellij.dlanguage.stubs.DLanguageModuleDeclarationStub
import java.io.IOException

class ModuleDeclarationStubElementType(debugName: String) :
    DNamedStubElementType<DLanguageModuleDeclarationStub, DLanguageModuleDeclaration>(debugName) {
    override fun createPsi(stub: DLanguageModuleDeclarationStub): DLanguageModuleDeclaration {
        return DlangModuleDeclarationImpl(stub, this)
    }

    override fun createStub(psi: DLanguageModuleDeclaration, parentStub: StubElement<*>?): DLanguageModuleDeclarationStub {
        return DLanguageModuleDeclarationStub(parentStub, this, psi.name, psi.attributes)
    }

    @Throws(IOException::class)
    override fun serialize(stub: DLanguageModuleDeclarationStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.name)
        stub.attributes.write(dataStream)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>?): DLanguageModuleDeclarationStub {
        return DLanguageModuleDeclarationStub(
            parentStub, this, dataStream.readName(),
            read(dataStream)
        )
    }
}
